### PR TITLE
Node-RED multi-arch Node v10 Docker images [amd64, arm32v6, arm32v7, arm64v8]

### PR DESCRIPTION
**Pull request by RaymondMouthaan**
Hi all,

This PR adds multi-architecture (amd64, arm32v6, arm32v7 and arm64v8) support for node-red docker images and is manifest listed. The base image is node v10. This PR is the successor of PR #114. 

As i wrote earlier ...

> @knolleary,
> @dceejay, thanks for your feedback :)
> 
> It has been a while, but meanwhile I have been working on a new PR regarding Node-RED Docker images.
> 
> In a nutshell the new PR will cover the following:
> 
> * Multi-arch support for arm32v6, arm32v7, arm64v8, amd64
> * Node 10 (dropped Node 6 and 8)
> * Python 2 and Python 3
> * Node-RED build-in RPI-GPIO support
> * Keep the images as small as possible (preferable Alpine based)
> 
> Currently all images and manifest lists are tested on best effort, this means tested if the right images are pulled and started successfully. arm32v6, arm32v7 are tested on a Raspberry PI 3B and amd64 on a Mac mini. Arm64v8 has been tested, thanks to @PedroLamas 👍 
> 
> I am updating the README, as well taking into account the comments you made above.
> 
> [Here](https://cloud.docker.com/repository/docker/raymondmm/node-red/tags) you find all Docker images and the builds can be found at [Travis-CI](https://travis-ci.org/RaymondMouthaan/node-red-docker), feel free to test and give feedback on the matter.

Best regards,
Ray

### CODE REVIEW

File: .docker/Dockerfile.alpine [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ARG BUILD_REF
+ARG NODE_RED_VERSION
+ARG ARCH
+ARG PYTHON_VERSION
+ARG QEMU_ARCH
+ARG OS
+
+LABEL org.label-schema.build-date=${BUILD_DATE} \
+    org.label-schema.docker.dockerfile=".docker/Dockerfile.alpine" \
+    org.label-schema.license="GNU" \
```
Comment (knolleary) [inline .docker/Dockerfile.alpine]: What license is this? All of Node-RED project assets are Apache-2.0

(Same comment on all of the Dockerfiles)

File: .travis.yml [hunk 2]
```diff
@@ -1,49 +1,92 @@
...
+      ./.docker/docker.sh push
+
+      # Docker Logout
+      docker logout
     fi
+
+jobs:
+    include:
+        - stage: manifest
+          if: tag =~ ^v
```
Comment (knolleary) [inline .travis.yml]: Am I right to read this as meaning it will only push to dockerhub when we push a tag?

**Comment by diversit** [comment]
Since Node v8 support is being dropped, does this PR then deprecate the PR's #113, #111, #98 and #28? (so those can be rejected)

**Comment by knolleary** [comment]
@diversit yes, this PR will address a lot of the open issues and PRs

File: .docker/Dockerfile.alpine (hunk 1 above)
Comment (RaymondMouthaan) [reply .docker/Dockerfile.alpine]: I'll modify asap!

File: .travis.yml (hunk 2 above)
Comment (RaymondMouthaan) [reply .travis.yml]: I'll explain later today.

Comment (RaymondMouthaan) [reply .travis.yml]: @knolleary, i've added some comments to explain, your question.

**Comment by RaymondMouthaan** [comment]
@knolleary @dceejay

A number of open questions:

- [ ] Should we add _big fat_ images with dev tools included, tag them as `<node_red_version>-<os>-<arch>-devtools`?, If yes, which dev tools should be included?

- [ ] Should we add Node v12 based images?

- [ ] In current PR some modifications are applied to the build-in RPI-GPIO support, to make it work for Python 3, or completely removed for non-arm32 architectures. Would it be feasible to completely remove them from Node-RED core and point users to the more elegant approach [node-red-node-pi-gpiod](https://github.com/node-red/node-red-nodes/tree/master/hardware/pigpiod)?

**Comment by fabnavigator** [comment]
First of all. Thank you so much for this image! I was dead in the water without it.

I have two node-red containers. In the first case, Portainer is telling me that the container is unhealthy, but I didn't add a health check. Below is my compose text. Could it have something to do with my non-standard port? I didn't see this issue with the nodered/node-red-docker:0.20.7-rpi image. Besides being unhealthy, my flows seem to be working.

node-red-bose:
container_name: node-red-bose
image: raymondmm/node-red:latest-rpi
# image: nodered/node-red-docker:0.20.7-rpi
user: '1001'
ports:
- 1881:1881
volumes:
- /home/fabnav/opt/node-red-bose:/data
- /etc/localtime:/etc/localtime:ro

Below is the log file:
,

node-red-docker@0.20.7 start /usr/src/node-red,
node $NODE_OPTIONS node_modules/node-red/red.js -v $FLOWS "--userDir" "/data",
,
18 Aug 12:54:34 - [info] ,
,
Welcome to Node-RED,
===================,
,
18 Aug 12:54:34 - [info] Node-RED version: v0.20.7,
18 Aug 12:54:34 - [info] Node.js version: v10.16.2,
18 Aug 12:54:34 - [info] Linux 4.19.58-v7+ arm LE,
18 Aug 12:54:35 - [info] Loading palette nodes,
18 Aug 12:54:38 - [info] Dashboard version 2.16.0 started at /ui,
18 Aug 12:54:39 - [info] Settings file : /data/settings.js,
18 Aug 12:54:39 - [info] Context store : 'default' [module=localfilesystem],
18 Aug 12:54:39 - [info] Context store : 'memory' [module=memory],
18 Aug 12:54:39 - [info] User directory : /data,
18 Aug 12:54:39 - [warn] Projects disabled : editorTheme.projects.enabled=false,
18 Aug 12:54:39 - [info] Flows file : /data/flows.json,
18 Aug 12:54:39 - [info] Server now running at http://127.0.0.1:1881/,
18 Aug 12:54:39 - [info] Starting flows,
18 Aug 12:54:39 - [info] Started flows,

My second container doesn't use a non-standard port, and isn't unhealthy. It uses timers though, and doesn't seem to be picking up the correct time zone, as the nodered/node-red-docker:0.20.7-rpi image did. Below is the compose text.

node-red:
container_name: node-red
image: raymondmm/node-red:latest-rpi
# image: nodered/node-red-docker:0.20.7-rpi
depends_on:
- postgres
- home-assistant
user: '1001'
ports:
- 1880:1880
volumes:
- /home/fabnav/opt/node-red:/data
- /etc/localtime:/etc/localtime:ro

Below is the log file:
,

node-red-docker@0.20.7 start /usr/src/node-red,
node $NODE_OPTIONS node_modules/node-red/red.js -v $FLOWS "--userDir" "/data",
,
18 Aug 13:18:35 - [info] ,
,
Welcome to Node-RED,
===================,
,
18 Aug 13:18:35 - [info] Node-RED version: v0.20.7,
18 Aug 13:18:35 - [info] Node.js version: v10.16.2,
18 Aug 13:18:35 - [info] Linux 4.19.58-v7+ arm LE,
18 Aug 13:18:36 - [info] Loading palette nodes,
18 Aug 13:18:41 - [info] Dashboard version 2.16.0 started at /ui,
18 Aug 13:18:41 - [info] Settings file : /data/settings.js,
18 Aug 13:18:41 - [info] Context store : 'default' [module=localfilesystem],
18 Aug 13:18:41 - [info] Context store : 'memory' [module=memory],
18 Aug 13:18:41 - [info] User directory : /data,
18 Aug 13:18:41 - [warn] Projects disabled : editorTheme.projects.enabled=false,
18 Aug 13:18:41 - [info] Flows file : /data/flows.json,
18 Aug 13:18:41 - [info] Server now running at http://127.0.0.1:1880/,
18 Aug 13:18:42 - [info] Starting flows,
18 Aug 13:18:42 - [info] [inject:18:00 on] crontab = 00 18 * * *,
18 Aug 13:18:42 - [info] [inject:20:00 on] crontab = 00 20 * * *,
18 Aug 13:18:42 - [info] [inject:01:30 off] crontab = 30 01 * * *,
18 Aug 13:18:42 - [info] [inject:6:00] crontab = 00 06 * * *,
18 Aug 13:18:42 - [info] [inject:5:00] crontab = 00 05 * * *,
18 Aug 13:18:42 - [info] [inject:9:00 on] crontab = 00 09 * * *,
18 Aug 13:18:42 - [info] [inject:daily initialization to false] crontab = 00 18 * * *,
18 Aug 13:18:42 - [info] [inject:19:00] crontab = 00 19 * * *,
18 Aug 13:18:42 - [info] [inject:22:00] crontab = 00 22 * * *,
18 Aug 13:18:43 - [info] [inject:midnight] crontab = 01 00 * * *,
18 Aug 13:18:43 - [info] Started flows,
18 Aug 13:18:43 - [info] [server:Home Assistant] WebSocket Connected to http://home-assistant:8123,

Thanks,

Bill

**Comment by knolleary** [comment]
Please raise an issue against Raymonds repo so he can help you there. Adding your question to a pull request is not the right thing to do.

**Comment by fabnavigator** [comment]
I couldn't find it, but I'll look some more. At least I didn't create another issue. That's progress.
P.S. I'm a great fan of your work!

**Comment by RaymondMouthaan** [comment]
> I couldn't find it, but I'll look some more. At least I didn't create another issue. That's progress.
> P.S. I'm a great fan of your work!

-> https://github.com/RaymondMouthaan/node-red-docker/issues

**Comment by xalexslx** [comment]
Hey @RaymondMouthaan 

Are we waiting something I can help test on this PR so it be merged?

Can I suggest split the open questions on future PR's so we don't hold this too much? I came from HA community and some nodes (eg chatbot) already updated requiring newest node version. I discovered it while redoing my setup. Therefore, I'm waiting this PR to be able to use the rpi image again.

Regards,

**Comment by knolleary** [comment]
@xalexslx we're working on it. I hope to get it merged this week - but please understand this is one of many tasks we're working on to get to Node-RED 1.0. And the sheer scale of the changes this PR introduces means it cannot be rushed.

I had asked some questions in my view of this PR and I say Raymond has responded. I have not had time to review his responses yet - but will do tomorrow.

You are right about those questions (node 12, full-fat images) - they are further enhancements that we can make in later updates. 

We'd certainly welcome any testing you wanted to do of the changes in this PR.

**Comment by RaymondMouthaan** [comment]
@xalexslx, if you want to test before the merge - you can find my images [here](https://cloud.docker.com/u/raymondmm/repository/docker/raymondmm/node-red) .. if you find some problems, please raise an issue [here](https://github.com/RaymondMouthaan/node-red-docker/issues). Also note the README, a few things might have changed ...

**Comment by RaymondMouthaan** [comment]
@knolleary, 

Please **don't** merge this PR, currently I am still working on it :) 

Subjects I am working are:

[√] Replace Node buster-slim arm32v7 images with Node Alpine arm32v7, so all Node-RED images are Alpine Linux based.
[√] Node 10 images and Node 12 images
[   ] Devtools images
[   ] [bcryptjs](https://www.npmjs.com/package/bcrypt) (required to secure Node-RED)
[   ] Python 2.x and Python 3.x
[   ] Tagging all images and manifest lists
[   ] Update README accordingly 

As soon as all of these are done, I'll raise new PR. For now I'll close this one.

Best regards,
[Raymond](https://github.com/RaymondMouthaan/node-red-docker)

**Comment by RaymondMouthaan** [comment]
New PR can be found at #127

**Pull request closed without merging**